### ISSUE TITLE
No support for very high resolution images

### ISSUE THREAD

**Issue by t-pearce** [Mavichist/Fountain#8]
Seems you're limited to 8192px across, which may not be enough for certain high resolution maps; the map I was going to attempt to render would be 256km^2/pixel. The inclusion of some kind of modular terrain system would improve the tool no end.

**Comment by t-pearce** [comment]
Performance at maximum resolution is also very poor; with no brush strokes, effects, or gradients, it takes upwards of half a second to scroll in or out.

**Comment by Mavichist** [comment]
Sampling from and rendering high-resolution images is obviously going to be a more costly operation than for lower resolutions. The 8192 limit is imposed in case the user hits an extra key by accident and ends up trying to create an image that occupies too much RAM and causes a crash. I didn't think it would realistically cause any problems because 8192x8192 is utterly enormous and for most applications; totally sufficient.

I'm not sure I know what you mean by "modular terrain system", but a definition would be kinda handy if you want a more detailed response. At the moment my response is; you can have as many renders as you want in a given document, so compartmentalize your bigger map into smaller parts that don't cause your lag problems.

**Comment by Mavichist** [comment]
Also; I just ran the math and I'm not really sure I understand why rendering at 256km^2/pixel is an issue. 256km^2 is an area that is 16kmx16km in size, so to render an image that encompasses the whole planet (roughly 40,000km in circumference) you would only need an image roughly 2500 pixels wide. This isn't even close to the 8192x8192 limit imposed by Fountain.

**Comment by t-pearce** [comment]
A more elegant solution to the high resolution memory could be to alert the user that they need approximately X GB of memory to be able to create an image that big, rather than outright disallow the usage.

It's not unusual for people using image editing software to work with images in the 7-8k^2 px range. Conceivably, Fountain could conceivably be used for images of such scale, and users familiar with other painting/art/illustration software may expect it to function well in that range, even if you didn't intend it to be. There's a certain element of the death of the developer (née author) here.

Regarding the 40,000km circumference planet, that is assuming that the map is best rendered in a square. Not all worlds created via this tool are necesssarily going to have equally sized edges. For instance, the world I had in mind is toroidal, and as such, has an approximate rectangular aspect ratio of 1:4. Further, terrain difference over 16km can be rather drastic. Large swathes of the grand canyon would have significant resolution loss at this size, for instance.

As for "modular terrain system", forgive my creation of jargon, I meant a system by which single map on the screen was rendered by knitting a series of sub-maps together. This way, you only need to render/memorise the submap being worked on, and the ones adjacent to it, but can allow technically arbitrarily sized maps. That is to say, a map could be 16 x 16 submaps, where each submap is 1024x1024px. If one were working on submap (3, 4), then submaps (2-4, 3-5) would be visible (3x3 grid), creating an active render area of 3072 x 3072 px, but the map as a whole would still have a total area of 16k x 16k px, at the price of 1GB of storage space on disk (assuming 4-channel colour). The right edge of submap (2, 3) would map to the left edge of submap (2, 4) for the purposes of brush strokes etc, etc.

This, of course, requires some storage access. However, longer loading times for larger discrete areas is generally preferable, at least in my eyes, than a stuttering load of continuous areas. Another issue comes where the brush approaches the edge of the adjacent submaps, but that could be solved by limiting the brush cursor from approaching within 0.5 * brush width/height. As such, the brush would never enter the 2nd order adjacent submaps. An overlay of a couple of pixels of inverted color could be used to denote the boundry.

As for rendering images under this modular system, downsampling would of course be needed for full renders of large maps, but this method also allows partial renders in higher resolutions, allowing greater detail.

If you really felt like going for it, you could also include upsampling from a smaller map; for each submap, divide equally along both axes n times to create n^2 equally sized and similarly proportioned sub-submaps, map each pixel to an n*n area, and define them as the new submaps. Detail can be added from there, as the user desires; the full render probably won't be affected as much as regional ones, but it allows for a degree of top-down design that some users are going to really appreciate.

Please forgive the wall of text. I'm passionate about such programs, even if my own attempts never really got very far. I'm also not very experienced in using github as a feedback toolx, so if this is an inappropriate place to make such suggestions, feel free to direct me to a better forum.

**Comment by Mavichist** [comment]
I thank you for your detail - walls of text are useful.

I'm happy to include a warning as opposed to a hard limitation, but you already recognize that not adhering to this limitation makes the program unwieldly. It's not accounted for in the use-case and was taken into consideration when imposing the limit.

I'm not sure I understand the sentiment that there's an element of 'death to the author' when it comes to software design. You're more or less saying that people should be able to use programs to do things that the programs were not designed to do, which seems like a contradiction. We impose limitations and abstain from accommodating extreme cases all the time because we're trying to follow a particular design path and deliver certain use-case. You seem to be suggesting that because a few people will want this feature, I have no choice but to implement it despite any detriment to my time.

I'm aware that for many people who use image editing applications; large images are commonplace. This, however, is not that use-case. Maps are often broken up into sub-regions for the sake of compartmentalization and readability. There is a reason Fountain was designed to accommodate many different renders at once; it's meant to be like an atlas. It was never designed under the assumption that you would be trying to render an entire planet in one image, with a level of detail that would otherwise be assigned to individual map segments.

Furthermore, and as a side note; you must also recognize that your case is made even more extraneous by the fact that you're attempting to render a torroidal planet. If I added any support for torroidal worlds, it would come at the end of a very long road of changes that are more relevant to more people. I don't mean to be dismissive or disrespectful, but I don't think many people using Fountain will be rendering this kind of map (though I am aware that there might be a higher percentage than otherwise anticipated because of the Y-axis wrapping).

As for the modular mapping - I'm at a crossroads on this issue. I've already incorporated a way to add more than one map component to a document and, though a blunt instrument, it will satisfy most end users. Your solution is one I've thought about before, but neglected to implement, not because it's a bad idea, but because it's an enormous overhaul of the rendering engine, brush system and underlying data structures that only really serves the purposes of a niche group of people. It's not a change that I've ruled out, so i'll keep this issue open for further discussion, but you should know that there are patches and features that come in at a far higher priority at the moment because they affect many more people, and will consume much less of my time.

Your last technical commentary seems to be suggesting the use of an octree structure (or equivalent) to lend more detail to larger maps. Such a solution isn't out of reach, and I have an octree implementation that could be of use, but I must say that it may result in more harm than good. The logistics of painting across different borders and resolutions using the same brush sizes are incredibly complex, but beyond that; such a change would undermine the current document structure, rendering every current saved file redundant and causing a migraine for anyone using current versions.

As for GitHub; this is a fine place for this - it's more constructive than I first thought it would be. I won't close this issue until something is done, but I need you to know that it may be a while yet before anything is.

**Comment by t-pearce** [comment]
I'm going to try to respond to each of your paragraphs on a one-to-one basis, because this is getting perhaps a little off track.

I feel like having a warning in place of a hard limit, or at least some way of overriding the hard limit that doesn't involve going into the source code, would be better than just flat out replacing the number. Certainly, your solutions to the issue is a solution, but I feel like it's not the most elegant one available. You are, of course, free to disagree with that.

My sentiment regarding the death of the author isn't to say that people should be able to use programs to do what those programs aren't supposed to do, it's that they'll try to push the boundries of it. A classic case in point is Shockwave/Flash, which has ended up doing so much more than was originally intended. Whether or not you want to implement it or not is, of course, up to you. I'm sorry if I made it sound otherwise, but I wouldn't be suggesting if I didn't think it would be worth it to have.

Knowing that Fountain was designed to portray subsections of a map, rather than a planetary map, paints the design decisions in a different light. However, not having the ability to create a detailed overworld map at a good resolution is, at least in my opinion, a sorely lacking feature. I feel like the program would benefit greatly from it.

Toroidal planets actually map onto rectangles very nicely, the distortion isn't really much worse than a Mercator projections (I don't believe). Cut it across the circumfrence and straighten to form a cylinder, then cut along the length of the cylinder, and you've got a rectangle.

I entirely understand the prioritising of these features and fixes. I've been in the same situation myself. I do think an option to use that kind of mapping would be a great improvement to the tool, and I'd certainly make a huge amount of use of it were it implemented, but without that kind of tiling, I'm somewhat limited. Obviously, I'm aware that it's a massive edge case, but the depth it'd add to the program would be immense. Perhaps a v2 release?

An octree would be very similar (an n^2-tree, perhaps, in this case? I'm not too hot on the official jargon for these things) to the proposed system, yes. While it may cause more harm than good to the current implementation, and I'd guess without intimate knowledge that it'd require a large overhaul of the system as it stands. However, I don't believe that it'd render the current document structure useless. For one, there could be two types of map, which use different structures. For another, rendering the descendant or ancestor renderings of a region could be done on the fly, and the functionality could be disabled by default. I'm certain there'd be a way to marry the two, and if not, there'd always be the option to import the old save format via some kind of parser. 

Of course, whether or not you're willing or able to commit the time to the project to make it happen, is another thing entirely, and indeed none of my business. I'm sure you have other projects, work, family, friends, etc, who'd like to see you every now and again. That said, if you think it's of value and you do have the time, I'd be happy to discuss it further.

**Comment by Mavichist** [comment]
I don't agree that the inability to output and handle large render sizes is 'sorely lacking'. Fountain isn't designed to be a complete suite and I regularly use it in conjunction with other software to create a finished product. It's designed to do a very specific thing; give people the ability to paint using custom functions and manipulate topography, and no more than that. I see the ability to output enormous renders as partially re-inventing the wheel because you can do that in Gimp (or equivalent) by stitching together the regional renders. I agree that outputting a whole render would be nice, but it's really not something that I can justify dedicating myself to at this point in development because the ability to do so is present elsewhere.

As for the octree implementation distorting the file format; I'm trying to change the file format at the moment so it's more extensible. That may help, but the issue is the changes to the underlying heightmap data structures. If I fiddle with those; things will become incompatible very quickly. Beyond that; it's an option, though I'd opt for some form of modular system that extends the current document so I can salvage peoples current projects a little safer. Alternate parsing is going to be necessary anyway, but the cleaner it is; the better.

**Comment by t-pearce** [comment]
Are you saying that Fountain isn't designed to be a complete suite **now**, or that it will never be a complete suite?

**Comment by Mavichist** [comment]
I'm not trying to tell you that I'm content to leave it half-baked. I didn't mean for it to sound like that.

Certainly it didn't start out with the intention of being a one-stop-shop though. I had a very clear design directive in mind for it when I started, and I'm trying to maintain that without getting sidetracked.

I have no intentions of delivering support for things like text overlay, annotations and symbols, as an example. It's purely for topology. It's not designed to do everything - there'd be no point. Photoshop and Gimp do that kind of thing so much more thoroughly than I could ever hope to. I'd rather focus on delivering something that lets you make landscapes with a high degree of flexibility, and specialize the program so that it does that well. To my mind at least; there's absolutely no point reinventing the wheel and detracting from the functional niche that the software fills.

**Comment by t-pearce** [comment]
I can definitely see an argument for not supporting text overlay, annotations, and symbols (though they would be a great addition if you decided to change your mind!), but lacking support for those high res images or knitting multiple images together means that those who're hopeless with the more nuanced usage of Gimp et al are relegated to single-image maps. For instance, I could happily import an image into Gimp and add some text/icons, but I wouldn't even know how to start to blend together two disparate maps made in Fountain that might not necessarily line up cleanly. Again, I wouldn't be making such a persistent case if I didn't think there was reason to, but it sounds like you disagree in that regard.

**Comment by Mavichist** [comment]
I said in one of my previous replies that the ability to stitch the maps together would be a nice feature. I intend to add it at some stage. For now though; I really think it's a luxury best left to other programs that do it better. At this early stage, I'm not convinced I should be spending so much of my time developing niche features.

As for the complexity issue; I really think that this program is harder to learn than Gimp or Photoshop. It's meant to be; the scripting and extensibility is what gives it its appeal and its power, but it's definitely paid for in learning curve. I think that if you can get a handle on Fountain to the point where you have a map spanning multiple renders; you can get a handle on how to match up map segments with other image editors.

**Comment by t-pearce** [comment]
I suppose that's true for the general population, yes. I come from a programming background myself, with experience in the language and similar functions already, so making Fountain work is a lot easier than Gimp or Photoshop. Sorry if I'm not keeping best track of what's been said previous.

**Comment by Mavichist** [comment]
It's been a bit of an Olympics - I've been getting lost a fair bit. I didn't mean that in a passive-aggressive fashion.

Honestly it's been a pleasure to talk to you about it, even though we don't always agree and we're coming from two different perspectives. It's refreshing to speak with someone who's as passionate and knowledgeable about these things as I am, if not more so. I do apologize if I seemed crass in the beginning. When the thread started and I read you talking about 'taking half a second to scroll' I thought your criticism was going to be superficial and petty, but as usual I read far too much into it and I was very, very wrong. You've had some truly insightful things to say, and in great technical detail, and I thank you for putting up with my brashness in order to say them.

**Comment by t-pearce** [comment]
Not at all, it's been a (somewhat tonally confusing) pleasure to talk to you about it too. I don't know how often you check it, but I believe I've reached out to you on reddit if you want to bounce ideas off me or hear a few of my other thoughts on it. I'd post them as an issue, but if I were being true to my catagorising self, they'd each have their own issue and probably make you feel like you've got a mountain of work, and on top of that, I don't know what other people've suggested.
